aten/src/ATen/native/TensorShape.cpp

```diff
@@ -392,6 +390,90 @@ Tensor cat(TensorList tensors, int64_t dim) {
   return result;
 }
 
+Tensor block_diag(TensorList tensors) {
+  Tensor result;
+  if (tensors.size() == 0) {
+    result = at::empty({0});
+    return result;
+  }
+
+  // Use at::result_type to find the right tensor to copy options from, so
+  // that the output data type is correct
+  const Tensor* options_tensor = &tensors[0];
+  ScalarType output_scalar_type = options_tensor->scalar_type();
+  const Device& device = tensors[0].device();
+
+  for (size_t tensor_idx = 1; tensor_idx < tensors.size(); tensor_idx++) {
+    const Tensor* other_tensor = &tensors[tensor_idx];
+
+    TORCH_CHECK(
+      other_tensor->device() == device,
+      "torch.block_diag: input tensors must all be on the same device.",
+      " Input 0 is on device ", device,
+      " and input ", tensor_idx, " is on device ", other_tensor->device()
+    );
+
+    ScalarType scalar_type = at::result_type(*options_tensor, *other_tensor);
+
+    if (scalar_type != output_scalar_type) {
+      output_scalar_type = scalar_type;
+      options_tensor = other_tensor;
+    }
+  }
+
+  int64_t result_dim0 = 0;
+  int64_t result_dim1 = 0;
+  std::vector<Tensor> tensors_2D(tensors.size());
+
+  // Sum the dimensions of the tensors, check tensor sizes,
+  // and expand all 0-D and 1-D tensors so that everything
+  // is 2-D
+  for (size_t tensor_idx = 0; tensor_idx < tensors.size(); tensor_idx++) {
+    const Tensor& tensor = tensors[tensor_idx];
+    int64_t ndims = tensor.dim();
+    TORCH_CHECK(ndims <= 2, "torch.block_diag: input tensors must have 2 or fewer dimensions");
```

Comment:
nit: We could make the error message clearer by stating the dimensionality of the input tensor that was not the right index